library/core/tests/future.rs

```diff
@@ -64,6 +64,40 @@ fn test_join() {
     });
 }
 
+/// Tests that `join!(…)` behaves "like a function": evaluating its arguments
+/// before applying any of its own logic.
+///
+/// _e.g._, `join!(async_fn(&borrowed), …)` does not consume `borrowed`;
+/// and `join!(opt_fut?, …)` does let that `?` refer to the callsite scope.
+mod test_join_function_like_value_arg_semantics {
+    use super::*;
+
+    async fn async_fn(_: impl Sized) {}
+
+    // no need to _run_ this test, just to compile it.
+    fn _join_does_not_unnecessarily_move_mentioned_bindings() {
+        let not_copy = vec![()];
+        let _ = join!(async_fn(&not_copy)); // should not move `not_copy`
+        let _ = not_copy; // OK
```

Comment:
Hmm, I've just tested this test suite against the current state of `master`, and this one test fails to check what I wanted it to check: the `let _ = not_copy;` does not count as an access to `not_copy`, and thus its is allowed even when moved out. We need to append a `&` to avoid that:

```suggestion
        let _ = &not_copy; // OK
```

cc @joshtriplett: you've already `r+` the current head of the branch, but I'd like to have this one commit applied and approved. If it seems sensible to you, then you can click on `Apply suggestion` yourself and re `r+`

FWIW, the two other tests, and this one once it has the `&`, do all fail under the current `master` 🙂